### PR TITLE
Generate TChannel endpoints

### PR DESCRIPTION
**Pull request by ChuntaoLu**

### CODE REVIEW

File: examples/example-gateway/endpoints/baz/call_tchannel.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+	"endpointType": "tchannel",
+	"endpointId": "baz",
+	"handleId": "call_tchannel",
+	"thriftFile": "endpoints/baz/baz.thrift",
+	"thriftFileSha": "{{placeholder}}",
+	"thriftMethodName": "SimpleService::Call",
+    "workflowType": "tchannelClient",
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/call_tchannel.json]: should be custom workflow here with no clientName or clientMethod.

File: examples/example-gateway/endpoints/baz/call_tchannel.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	"clientName": "Baz",
+	"clientMethod": "Call",
+	"testFixtures": [],
+	"middlewares": [],
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/call_tchannel.json]: TChannel does not have middlewares or test fixtures so lets remove those fields.

File: examples/example-gateway/endpoints/baz/call_tchannel.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	"reqHeaderMap": {},
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/call_tchannel.json]: `reqHeaderMap` and `resHeaderMap` are only read by the workflow, so not needed here.

File: codegen/templates/tchannel_endpoint.tmpl [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+// Handle handles RPC call of "{{.ThriftService}}::{{.Name}}".
+func (h *{{$handlerName}}) Handle(
+	ctx context.Context,
+	reqHeaders map[string]string,
+	wireValue *wire.Value,
+) (bool, zanzibar.RWTStruct, map[string]string, error) {
+	var res {{$genCodePkg}}.{{.ThriftService}}_{{.Name}}_Result
+
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: We want to add header validation here.

We have a `zanzibar.http.reqHeaders` annotation that we want to respect.

File: codegen/templates/tchannel_endpoint.tmpl [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+		Clients: h.Clients,
+		Logger: h.Logger,
+	}
+
+	{{if and (eq .RequestType "") (eq .ResponseType "")}}
+	respHeaders, err := workflow.Handle(ctx, headers)
+	{{else if eq .RequestType ""}}
+	r, respHeaders, err := workflow.Handle(ctx, reqHeaders)
+	{{else if eq .ResponseType ""}}
+	respHeaders, err := workflow.Handle(ctx, reqHeaders, &req)
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: `reqHeaders` here should implement `zanzibar.Headers` interface instead of `map[string]string`

Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: respHeaders should also be `zanzibar.Header`

File: codegen/templates/tchannel_endpoint.tmpl [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	{{else}}
+	r, respHeaders, err := workflow.Handle(ctx, reqHeaders, &req)
+	{{end -}}
+
+	{{if eq (len .Exceptions) 0 -}}
+		if err != nil {
+			return false, nil, nil, err
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: want to return `respHeaders` here.

File: codegen/templates/tchannel_endpoint.tmpl [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+		}
+		res.Success = r
+	{{else -}}
+		if err != nil {
+			switch v := err.(type) {
+			{{$method := .Name -}}
+			{{range .Exceptions -}}
+				case *{{$genCodePkg}}.{{.Name}}:
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: `{{.Type}}` here.

File: codegen/templates/tchannel_endpoint.tmpl [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+					if v == nil {
+						return false, nil, nil, errors.New(
+							"Handler for {{$method}} returned non-nil error type *{{.Name}} but nil value",
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: `{{.Type}}`

File: codegen/templates/tchannel_endpoint.tmpl [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+						)
+					}
+					res.{{.Name}} = v
+			{{end -}}
+				default:
+					return false, nil, nil, err
```
Comment (Raynos) [inline codegen/templates/tchannel_endpoint.tmpl]: return `respHeaders`

**Comment by Raynos** [comment]
Forgot to mention, we dont need to generate a proxy workflow for tchannel servers.

Let's refactor this example to be a custom tchannel workflow and remove the workflow generation code for now.

**Comment by ChuntaoLu** [comment]
@Raynos addressed the comments, please review again.

File: examples/example-gateway/build/endpoints/baz/baz_simpleservice_method_call_tchannel.go [hunk 10]
```diff
@@ -56,89 +56,44 @@ type SimpleServiceCallHandler struct {
...
 		Clients: h.Clients,
 		Logger:  h.Logger,
 	}
 
-	respHeaders, err := workflow.Handle(ctx, reqHeaders, &req)
+	wfRespHeaders, err := workflow.Handle(ctx, wfReqHeaders, &req)
+
+	respHeaders := map[string]string{}
+	for _, key := range wfRespHeaders.Keys() {
+		respHeaders[key], _ = wfReqHeaders.Get(key)
```
Comment (Raynos) [inline examples/example-gateway/build/endpoints/baz/baz_simpleservice_method_call_tchannel.go]: bug: copy over wfResHeaders, not wfReqHeaders.

File: runtime/server_header.go [hunk 11]
```diff
@@ -83,3 +82,57 @@ func (zh ServerHTTPHeader) Keys() []string {
...
+// on map[string]string
+type ServerTChannelHeader map[string]string
+
+// Get retrieves the string value stored on a given header. Bool
+// return value is used to distinguish between the presence of a
+// header with golang's zerovalue string and the absence of the string.
+func (th ServerTChannelHeader) Get(
+	key string,
+) (string, bool) {
+	// TODO: Canonicalize strings before lookup.
```
Comment (Raynos) [inline runtime/server_header.go]: I'd rather not do that. The correct string literal should be used here with the correct casing.

File: examples/example-gateway/endpoints/baz/baz_call_tchan.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	clientRespHeaders, err := w.Clients.Baz.Call(ctx, clientReqHeaders, clientReq)
+	respHeaders := zanzibar.ServerTChannelHeader(clientRespHeaders)
+	if err != nil {
+		switch v := err.(type) {
+		case *clientBaz.AuthErr:
+			return respHeaders, (*endpointBaz.AuthErr)(v)
+		default:
+			w.Logger.Error("baz.Call returned unexpected error")
```
Comment (Raynos) [inline examples/example-gateway/endpoints/baz/baz_call_tchan.go]: please log the error.

File: runtime/server_header.go (hunk 11 above)
Comment (ChuntaoLu) [inline runtime/server_header.go]: got it, copy pasted from the http ones

File: examples/example-gateway/endpoints/baz/baz_call_tchan.go (hunk 12 above)
Comment (ChuntaoLu) [inline examples/example-gateway/endpoints/baz/baz_call_tchan.go]: Done

File: examples/example-gateway/build/endpoints/baz/baz_simpleservice_method_call_tchannel.go (hunk 10 above)
Comment (ChuntaoLu) [inline examples/example-gateway/build/endpoints/baz/baz_simpleservice_method_call_tchannel.go]: Done, will have proper tests in future PRs to catch bugs like this.

**Pull request merged**